### PR TITLE
Feature: Find files in repo

### LINKED ISSUE #12638: Go to file functionality
Add functionality to quickly go a file by typing its filename.

### PR DESCRIPTION
**Pull request by rogerluo410**
Find files for the specified branch in repo.

Issue for the feature: 
[](url)https://github.com/go-gitea/gitea/issues/766

find-files

### CODE REVIEW

File: web_src/js/index.js
Comment (silverwind) [inline web_src/js/index.js]: Maybe move this to separate file `web_src/js/features/find.js`.

File: web_src/js/index.js
Comment (silverwind) [inline web_src/js/index.js]: Can you use https://api.jquery.com/wrap/ instead of `.html` (which is generally better to be avoided where possible for security reasons).

File: web_src/js/index.js
Comment (silverwind) [inline web_src/js/index.js]: I think you can use a [`Set`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) instead of a Array for better performance, unless there are duplicates expected.

File: web_src/js/index.js
Comment (rogerluo410) [reply web_src/js/index.js]: Created!

File: web_src/js/index.js
Comment (rogerluo410) [reply web_src/js/index.js]: Hi silverwind, 

Thanks for your suggestion. But, I found .Wrap() is different to .html(), which is to wrap element as same as .appendTo(), not going to convert string to html.

Also, the strings in html is from my own code, not from untrusted sources. So, there is no any security reasons for .html.

And there are many .html used in web_src/js/index.js...

File: web_src/js/index.js
Comment (rogerluo410) [reply web_src/js/index.js]: Hi, silverwind

Thanks for your suggestion. 

There isn't too much Insert or Remove operation in the array set.  It just stores the indexes of hit chars in the string. I think we no need to ask for higher performance...

Also, I didn't see anywhere use Set in the web_src/js/index.js

**Comment by lafriks** [comment]
Imho it will be quite performance killer for large repositories, imho filtering should be done server side using ui something like github

File: web_src/js/index.js
Comment (silverwind) [reply web_src/js/index.js]: My point is that if you're only checking for existance later on  via `.includes`, it's better to just use a Set because Set existance check is O(1) complexity while the same for an array is O(n), so will always be faster unless array length is 1.

File: web_src/js/index.js
Comment (silverwind) [reply web_src/js/index.js]: Yeah, it does not matter in this case as the wrapped HTML is already rendered before by the server. I just generally prefer to reduce `.html` usage.

File: options/locale/locale_en-US.ini
Comment (bagasme) [inline options/locale/locale_en-US.ini]: Why do you name it `find_file` when do you mean `Go to file`?

**Comment by KN4CK3R** [comment]
> 
> 
> Imho it will be quite performance killer for large repositories, imho filtering should be done server side using ui something like github

Github doesn't filter server side. On pageload there is a JSON request which returns all files of the repository. I noticed that the `vendor` directory is missing from the list but it doesn't look like they are using the `.ignore` file for this.

**Comment by silverwind** [comment]
> I noticed that the vendor directory is missing from the list but it doesn't look like they are using the .ignore file for this.

It's probably the linguist attributes:

https://github.com/go-gitea/gitea/blob/54d7b0f8e737cda864af981b29dda5116fc5f9ca/.gitattributes#L2-L3

Not sure how far we are for Gitea for the `.gitattributes` parsing but it'd certainly be nice if both it and maybe also `.ignore` could be applied in this search. But probably not worth to add it now if the infrastructure to parse such files is not present yet.

**Comment by KN4CK3R** [comment]
In https://github.com/go-gitea/lgtm the `vendor` directory is stripped too without `.gitattributes` or something. Maybe they have just a list of common excludes. I have created a `vendor` directory in a test repo and that was stripped too.

**Comment by silverwind** [comment]
Ugh, yeah if it's a hardcoded list, that's not going to be obvious. I suggest that if we do filtering based on files, either use `.gitattributes` or `.ignore`, ideally both.

**Comment by lafriks** [comment]
we can easily use go-enry there is method `enry.IsVendor("path/to/filename")` to filter out them

**Comment by lafriks** [comment]
> > Imho it will be quite performance killer for large repositories, imho filtering should be done server side using ui something like github
> 
> Github doesn't filter server side. On pageload there is a JSON request which returns all files of the repository. I noticed that the `vendor` directory is missing from the list but it doesn't look like they are using the `.ignore` file for this.

That's sounds more reasonable, as rendering all file list in html will be quite an overkill

File: routers/repo/find.go
Comment (lunny) [inline routers/repo/find.go]: 2021

**Comment by lunny** [comment]
I think maybe the solution may have better performance.

On find file page, when loading, we can get all entries of currently tree via ajax and filter on frontend.

When input a `/`, we can load all entires below this trees and filter on frontend.

All entries will be cached on this page, so that if you delete a `/`, then no data loading is necessary.

**Comment by rogerluo410** [comment]
> In https://github.com/go-gitea/lgtm the `vendor` directory is stripped too without `.gitattributes` or something. Maybe they have just a list of common excludes. I have created a `vendor` directory in a test repo and that was stripped too.
> 
> Edit: https://docs.github.com/en/github/searching-for-information-on-github/finding-files-on-github


> > I noticed that the vendor directory is missing from the list but it doesn't look like they are using the .ignore file for this.
> 
> It's probably the linguist attributes:
> 
> https://github.com/go-gitea/gitea/blob/54d7b0f8e737cda864af981b29dda5116fc5f9ca/.gitattributes#L2-L3
> 
> Not sure how far we are for Gitea for the `.gitattributes` parsing but it'd certainly be nice if both it and maybe also `.ignore` could be applied in this search. But probably not worth to add it now if the infrastructure to parse such files is not present yet.


I didn't think much about excluded directories.  That is really a problem.

So, what is the best solution of it? 

If repo has  `.gitattributes` or `.ignore` exists,  we use them to exclude directories,  otherwise, use hardcoded list to exclude instead ??

File: options/locale/locale_en-US.ini
Comment (rogerluo410) [reply options/locale/locale_en-US.ini]: Learn from Github ... Perhaps we can change a name for Gitea style :)

**Comment by silverwind** [comment]
> If repo has .gitattributes or .ignore exists, we use them to exclude directories, otherwise, use hardcoded list to exclude instead ??

I think just supporting `.gitattributes` via go-enry should be good enough for a start.

**Comment by rogerluo410** [comment]
> > If repo has .gitattributes or .ignore exists, we use them to exclude directories, otherwise, use hardcoded list to exclude instead ??
> 
> I think just supporting `.gitattributes` via go-enry should be good enough for a start. Maybe it could be a feature request to go-enry to support `.ignore` as well for search ignores, e.g. `enry.IsSearchIgnored("path/to/filename")`.

Ok, I will solve it, thank you.

**Comment by rogerluo410** [comment]
Hi @silverwind , 

It seems .gitattributes is not yet supported in go-enry.  

FYI:
go-enry

The current solution uses hardcoded list instead.  Created a hardcoded list to exclude dirs. And created an API to fetch files in frontend. When searching files,  set a threshold to show top N searched files, threshold is 50 now.

Performance:
search-file

**Comment by lafriks** [comment]
enry.IsVendor already has "hardcored" paths (node_modules, etc) so that is what should be used currently like it is already done in gitea code search feature

File: routers/api/v1/repo/find_files.go
Comment (lafriks) [inline routers/api/v1/repo/find_files.go]: enry.IsVendor should be used here

**Review by lafriks — CHANGES_REQUESTED** [review]

File: routers/api/v1/repo/find_files.go
Comment (lafriks) [reply routers/api/v1/repo/find_files.go]: Looks like comment in wrong place but you get the idea :)

Comment (rogerluo410) [reply routers/api/v1/repo/find_files.go]: used

**Comment by zeripath** [comment]
Please resolve conflicts.

**Comment by rempas** [comment]
What is the status of this merge requests? Why the conflicts are not getting fixed?

**Comment by lunny** [comment]
Please resolve the conflicts.

**Comment by rempas** [comment]
> Please resolve the conflicts.

Out of curiosity, don't you have the rights to do it in OP's place?

**Comment by rogerluo410** [comment]
> Please resolve the conflicts.

Hi Lunny, conflicts solved,  it can be merged now , but CI is blocked on testing-amd64: open /drone/src/coverage.out: permission denied.  How to fix it?

**Comment by wxiaoguang** [comment]
> > Please resolve the conflicts.
> 
> Hi Lunny @lunny , conflicts solved, it can be merged now , but CI is blocked on testing-amd64: open /drone/src/coverage.out: permission denied. How to fix it?

Thank you for resolving the conflicts. The CI failure is not related and I restarted it.

I am new here (too), and I try to answer the question about "merge":

A Gitea PR requires at least 2 approvals from maintainer team to get merged, this PR hasn't got one yet.

And I can still find some problems in code:

* Inconsistent naming style
    * `id="new-pull-request"` vs `id="ownerName"` (we have guideline about frontend, these ID prefers keba-case)
    * `tr_wrap` vs `treeLink`
* Escaping problem: ``` `<b>${entry[i]}</b>` ```

I haven't read the code carefully, but if there are still problems, it doesn't seem can be merged soon.

**Comment by rogerluo410** [comment]
> > > Please resolve the conflicts.
> > 
> > 
> > Hi Lunny @lunny , conflicts solved, it can be merged now , but CI is blocked on testing-amd64: open /drone/src/coverage.out: permission denied. How to fix it?
> 
> Thank you for resolving the conflicts. The CI failure is not related and I restarted it.
> 
> I am new here (too), and I try to answer the question about "merge":
> 
> A Gitea PR requires at least 2 approvals from maintainer team to get merged, this PR hasn't got one yet.
> 
> And I can still find some problems in code:
> 
> * Inconsistent naming style
>   
>   * `id="new-pull-request"` vs `id="ownerName"` (we have guideline about [frontend](https://docs.gitea.io/en-us/guidelines-frontend/), these ID prefers keba-case)
>   * `tr_wrap` vs `treeLink`
> * Escaping problem: `` `<b>${entry[i]}</b>` ``
> 
> I haven't read the code carefully, but if there are still problems, it doesn't seem can be merged soon.

Hi, thanks for your comment, I will change id naming to kebab-case. 
Also, `<b>${entry[i]}</b> ` is a string which will be convert to html at the function generateTrWithHighlight, something like `<br>` html tag won't be escaped.

File: routers/api/v1/repo/find_files.go
Comment (lafriks) [inline routers/api/v1/repo/find_files.go]: Do we really need hardcoded exclude paths?

**Comment by wxiaoguang** [comment]
I have no question about the code style or framework usage (these parts can be improved later, together).

The only question is the design of `fetchRepoFiles`. If I understand correctly, it loads all files in the repo for every page view. If the repo is large (Gitea repo has around 6000 files, and there are a lot of much larger repos), will this become a performance problem?

File: templates/repo/home.tmpl
Comment (wxiaoguang) [inline templates/repo/home.tmpl]: `id="new-pull-request"` seems incorrect

File: routers/api/v1/repo/find_files.go
Comment (wxiaoguang) [reply routers/api/v1/repo/find_files.go]: Agree to remove the hard-coded paths

File: routers/api/v1/repo/find_files.go
Comment (wxiaoguang) [inline routers/api/v1/repo/find_files.go]: It seems not worth to make it as an API. Once it becomes an API, the changing will be very difficult.

(some recent PRs also refactored some AJAX requests from API to internal AJAX endpoint only)

If the AJAX request is only used by a page, it could be put under the `/org/repo/....` , just like what GitHub does (`https://github.com/go-gitea/gitea/tree-list/62c74d215b3ba5dc34841268c5b9e09bc4531804`)

File: routers/api/v1/repo/find_files.go
Comment (wxiaoguang) [inline routers/api/v1/repo/find_files.go]: Should the `excludedDirs` be removed then?

File: templates/repo/find/files.tmpl
Comment (wxiaoguang) [inline templates/repo/find/files.tmpl]: Could the `style` be replaced by some existing UI CSS classes?

File: templates/repo/find/files.tmpl
Comment (wxiaoguang) [inline templates/repo/find/files.tmpl]: comment, maybe to be removed 😊

File: web_src/js/features/find.js
Comment (wxiaoguang) [inline web_src/js/features/find.js]: Does it `break` too early?

If there are 200 files loaded, then the first 50 (threshold) are not hit, will it break and result in an empty list?

Comment (rogerluo410) [reply web_src/js/features/find.js]: > I have no question about the code style or framework usage (these parts can be improved later, together).

> The only question is the design of fetchRepoFiles. If I understand correctly, it loads all files in the repo for every page view. If the repo is large (Gitea repo has around 6000 files, and there are a lot of much larger repos), will this become a performance problem?

That what I was concerned, also others.  But, when I watched Github's behavior for find files, it returns all file paths within one API request, and filter files on frontend. 

image

I was persuaded with the action...   It may meet performance problem if the number of files is so huge.  
Maybe filtering with debouncing on backend is a solution. And, I was confused why Github didn't do like that.  
Do you have solution ?

Comment (rogerluo410) [reply web_src/js/features/find.js]: > Does it `break` too early?
> 
> If there are 200 files loaded, then the first 50 (threshold) are not hit, will it break and result in an empty list?

Sure,  I will fix it.

**Comment by rogerluo410** [comment]
Changed to red for hit character.  

image

**Comment by KN4CK3R** [comment]
> The only question is the design of `fetchRepoFiles`. If I understand correctly, it loads all files in the repo for every page view.

Just to clarify: The data gets only requested on the "Go to file" page not for EVERY page view.

File: web_src/js/index.js
Comment (wxiaoguang) [reply web_src/js/index.js]: This algorithm is sub-string matching, it's unlikely to use `Set` here. For example:

```
keys(input) = aaabbbccc
entry(to_match) = abc
```

then the result might be `[a]aa[b]bb[c]cc`

File: web_src/js/index.js
Comment (wxiaoguang) [reply web_src/js/index.js]: A range-based escaping algorithm should be used. For example:

```
keys = "the/directory"
entry = "thedir"
```

Then the result should be:  

```
<span class=red>${escape("the")}</span>
<span>${escape("/")}
<span class=red>${escape("dir")}</span>
<span>${escape("ectory")}</span>
```

Otherwise, if the filename contains `<br>` or some JS code, the viewer may be attacked by XSS.

**Comment by wxiaoguang** [comment]
@rogerluo410 if you do not mind, I can do some refactoring (edit on this PR), especially for the XSS part, I hope this feature can be merged into 1.17

**Comment by rogerluo410** [comment]
> @rogerluo410 if you do not mind, I can do some refactoring (edit on this PR), especially for the XSS part, I hope this feature can be merged into 1.17

Sure. Thank you

**Review by wxiaoguang — APPROVED** [review]
Almost done from my side.

File: templates/repo/home.tmpl
Comment (wxiaoguang) [inline templates/repo/home.tmpl:78]:
```suggestion
						<button class="ui compact basic button">{{.i18n.Tr "repo.find_file.go_to_file"}}</button>
```

File: routers/web/repo/treelist.go
Comment (lunny) [inline routers/web/repo/treelist.go:25]: There is a performance problem here.

Comment (wxiaoguang) [reply routers/web/repo/treelist.go:25]: That's what GitHub does at the moment. And see my comment.

image

Comment (lunny) [reply routers/web/repo/treelist.go:25]: I think maybe Github will generate a cache json file when commit pushed to the branch.

Comment (wxiaoguang) [reply routers/web/repo/treelist.go:25]: I think GitHub has its own methods to do the optimization, instead of generating the cache when a commit is pushed to the branch, because the URL `https://github.com/{org}/{repo}/tree-list/{commit-id}` can get the file list for any commit id.

About performance:
* on `gitea` repo, around 6000 files (MacBook SSD)
    * `git ls-tree -t -l -r main > /dev/null  0.03s user 0.02s system 93% cpu 0.055 total`
* on `gitea` repo, around 6000 files (VPS, virtual disk around 100MB/s, 512MB memory)
    * `git ls-tree -t -l -r main > /dev/null  0.03s user 0.01s system 63% cpu 0.057 total`
* on `linux` repo, around 80000 files (VPS, virtual disk around 100MB/s, 512MB memory)
    * `git ls-tree -t -l -r master > /dev/null  0.74s user 0.73s system 29% cpu 5.049 total`
    *  not too bad, it will be much faster on a SSD with more memory.

To introduce a cache, I think it's out of the scope of this PR, and in my mind the cache data had better to be put inside the bare git repository on disk (eg: `/data/git/repositories/org/repo.git/gitea-cache/xxxxxx`) because the in-memory cache doesn't fit the file list data well (the file list data is large, and it should have a very long expiration time).

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
@lafriks needs your review again.

File: routers/web/repo/treelist.go
Comment (delvh) [inline routers/web/repo/treelist.go]:
```suggestion
		ctx.ServerError("ListEntriesRecursive", err)
```

File: routers/web/repo/treelist.go
Comment (delvh) [inline routers/web/repo/treelist.go]: For performance reasons, I think
```suggestion
	files := make([]string, 0, len(entries))
```
would be far better.

File: routers/web/repo/treelist.go
Comment (delvh) [inline routers/web/repo/treelist.go:54]:
```suggestion
	return entry.IsDir() || entry.IsSubModule() || enry.IsVendor(entry.Name())
```
Why the IsVendor check?
Should vendored files not be searchable?

File: templates/repo/home.tmpl
Comment (delvh) [inline templates/repo/home.tmpl:78]: That is one approach.
I think another one could think of would be to display a magnifying glass icon ( and perhaps a tooltip).

File: templates/repo/find/files.tmpl
Comment (delvh) [inline templates/repo/find/files.tmpl]: Isn't there already the `hidden` styleclass for that?
```suggestion
		<div id="repo-find-file-no-result" class="hidden ui row center mt-5">
```
Of course, that should also be changed in the JS then...

File: routers/web/repo/find.go
Comment (delvh) [inline routers/web/repo/find.go:22]: a) Aren't these two switched around?
b) What about using `SourceLink` instead of `DataLink`? Personally, I don't know what could be present under a `DataLink` entry...

File: web_src/js/features/repo-findfile.js
Comment (delvh) [inline web_src/js/features/repo-findfile.js:11]: I think `filter` should probably be renamed to `filterString` because I thought at first you want to supply a custom filtering function, and not a string.

File: web_src/js/utils.js
Comment (delvh) [inline web_src/js/utils.js]: I have huge problems understanding the algorithm.
Please add a few comments to explain what you are doing where and why it leads to the result you want.

File: templates/repo/find/files.tmpl
Comment (delvh) [inline templates/repo/find/files.tmpl:9]:
```suggestion
				<input id="repo-file-find-input" type="text" autofocus data-url-source-link="{{.DataLink}}" data-url-tree-link="{{.TreeLink}}">
```

File: web_src/js/features/repo-findfile.js
Comment (delvh) [inline web_src/js/features/repo-findfile.js]: Again, I have comprehension problems for this block.
Please add one or more comments.

Comment (wxiaoguang) [reply web_src/js/features/repo-findfile.js]: There are comments: ` // return the sub-match result as an array:  [unmatched, matched, unmatched, matched, ...]` and unit tests.

`matchRes[even]` means unmatched, `matchRes[odd]` means matched. You can add comments you like.

File: routers/web/repo/treelist.go
Comment (lafriks) [reply routers/web/repo/treelist.go:54]: I think it makes sense to skip vendored files, like we also do that in search

File: web_src/js/features/repo-findfile.js
Comment (delvh) [reply web_src/js/features/repo-findfile.js]: Yes, but I still can't understand the steps that are taken in the algorithm.
I don' want to maintain algorithms where I don't know why which step was performed...

File: routers/web/repo/treelist.go
Comment (wxiaoguang) [reply routers/web/repo/treelist.go:54]: See the comment of  KN4CK3R  https://github.com/go-gitea/gitea/pull/15028#issuecomment-802723331

I do not want to shorten the code too aggressively, these `if` are clear and extensible.

File: web_src/js/features/repo-findfile.js
Comment (delvh) [reply web_src/js/features/repo-findfile.js]: Also, if it is clear where what is done, it is easier to find and fix bugs.

File: routers/web/repo/find.go
Comment (wxiaoguang) [reply routers/web/repo/find.go:22]: Source is also a general word, not too much difference from `data`.

`data` is the data used by this page.

File: web_src/js/features/repo-findfile.js
Comment (wxiaoguang) [reply web_src/js/features/repo-findfile.js:11]: No difference in my mind.

File: routers/web/repo/find.go
Comment (delvh) [reply routers/web/repo/find.go:22]: Yeah, I can somewhat live with it, but the `data-url-data-link` attribute is too much:
Both `data` and `url` are effectively duplicated (link~url).
Is there no better name for that?

File: web_src/js/features/repo-findfile.js
Comment (wxiaoguang) [reply web_src/js/features/repo-findfile.js]: But I think it's just a very basic string comparison algorithm, I do not know how to write more meaningful comments.

Please point out which line you do not understand.

File: templates/repo/home.tmpl
Comment (wxiaoguang) [reply templates/repo/home.tmpl:78]: This layout is just copied from other buttons alongside it.

Comment (delvh) [reply templates/repo/home.tmpl:78]: Yes, but unlike other buttons, this one has an icon that you typically expect for this action.
I think most people are familiar with searching by clicking on a magnifying glass button.

File: web_src/js/features/repo-findfile.js
Comment (delvh) [reply web_src/js/features/repo-findfile.js]: Yeah, right, I mislabeled the unclear boundaries:
I understand the outer loop, but I don't really see the effect of the inner loop:
What is it supposed to do, concatenate the different folders into sections?
But why is every odd section red?

Comment (delvh) [reply web_src/js/features/repo-findfile.js]: And why don't I see that in the video?

File: routers/web/repo/treelist.go
Comment (delvh) [reply routers/web/repo/treelist.go:54]: Okay, if you want to keep the `if`s...
I'd prefer it shortened, but I can live with the current approach as well.

File: templates/repo/find/files.tmpl
Comment (wxiaoguang) [reply templates/repo/find/files.tmpl]: I do not think there is a common `hidden` CSS class.  I do not think it will hide the elements.

Could you help to confirm if it works?

File: web_src/js/features/repo-findfile.js
Comment (wxiaoguang) [reply web_src/js/features/repo-findfile.js]: See the unit tests.

```
expect(strSubMatch('abc', 'a')).toEqual(['', 'a', 'bc']);
```

if the target file path is "abc", to search "a", then "a" (odd) is highlighted to read.

File: templates/repo/find/files.tmpl
Comment (delvh) [reply templates/repo/find/files.tmpl]: `ag 'class=".*hidden' templates/ | wc -l` at least gives 68 results.
Either most of them are incorrect, or there is a `hidden` class...

File: web_src/js/features/repo-findfile.js
Comment (wxiaoguang) [reply web_src/js/features/repo-findfile.js]: > And why don't I see that in the video?

It's not in the video, it's in the comment https://github.com/go-gitea/gitea/pull/15028#issuecomment-1141906557 . The video makes the matched part as `<b>`.

Comment (delvh) [reply web_src/js/features/repo-findfile.js]: Interesting. Where is the `red` in the video?
But there we have the comment I meant: I thought that the odd indices pointed at something else, and not at the matched parts of the string. And that is definitely not obvious. But yeah, I can definitely see that conclusion by looking at the test.

File: templates/repo/home.tmpl
Comment (wxiaoguang) [reply templates/repo/home.tmpl:78]: Texts are more clear. GitHub also only puts texts for "Go to file".

In Gitea, all buttons in this line are text-only, no icon.

File: routers/web/repo/find.go
Comment (wxiaoguang) [reply routers/web/repo/find.go:22]: I think it's just fine, not too bad.

**Review by delvh — APPROVED** [review]

File: web_src/js/features/repo-findfile.js
Comment (wxiaoguang) [reply web_src/js/features/repo-findfile.js]: I added more comments with examples.

File: templates/repo/find/files.tmpl
Comment (wxiaoguang) [reply templates/repo/find/files.tmpl]: Nope, in most cases, it doesn't work .........

And you must grep `class="[^"]*hidden[^"]*"`,  `.*` is incorrect ........

File: routers/web/repo/treelist.go
Comment (lafriks) [reply routers/web/repo/treelist.go:54]: Personally I would also prefer shortened

**Review by lafriks — APPROVED** [review]

File: templates/repo/find/files.tmpl
Comment (wxiaoguang) [reply templates/repo/find/files.tmpl]: Hmm..... I think we can use `<div id="repo-find-file-no-result" class="ui row center mt-5" hidden>`  instead, the `[hidden]` is defined in `semantic.css`, which should be safe.

Comment (wxiaoguang) [reply templates/repo/find/files.tmpl]: > `ag 'class=".*hidden' templates/ | wc -l` at least gives 68 results. Either most of them are incorrect, or there is a `hidden` class...

Good question, after reading the related code/document again, I opened a new issue:
* https://github.com/go-gitea/gitea/issues/19926

**Comment by rempas** [comment]
Finally, it is merged!!!! Personally thanks everyone that worked in the feature!!! You people are awesome!!! :heart:

**Pull request merged**
